### DISCUSSION TITLE
Not able to kill the process during build

### DISCUSSION THREAD

**Discussion by groenroos** [11ty/eleventy#4104] [Enhancement Queue]
### Operating system

macOS Sonoma 14.4.1

### Eleventy

3.0.0

### Describe the bug

While Eleventy is building the site, it's not possible to kill the process with `Ctrl-C`. Only after the build is complete, will the `Ctrl-C` take effect and Eleventy exits with code 1.

With a large site where build times are in the tens of seconds, this can get frustrating, especially when you already know the build will fail (e.g. because of missing `npm` dependencies when switching git branches).

### Reproduction steps

1. Start a build of a large site
2. Wait until the templates start processing (i.e. after any async data has been fetched)
3. Try to kill the process with `Ctrl-C`
4. The process will not die until after the build is complete

### Expected behavior

The build should be interrupted and the process should exit instantly.

**Comment by nhoizey** [comment]
This is indeed a new behavior I've noticed with Eleventy 3.

<kbd>Ctrl</kbd> + <kbd>C</kbd> doesn't kill the shell execution, I have to use `ps` to get the process id then `kill` it.

**Comment by nhoizey** [comment]
There are similar issues open in npm's repository.

For example:
- https://github.com/npm/cli/issues/5976

**Comment by zachleat** [comment]
Likely related to #3361 and attempting to stop everything gracefully (shipped with v3.0.0)

**Comment by zachleat** [comment]
There _was_ a bug in this behavior that I’ve filed #3873 (but it doesn’t accurately reflect your issue).

As mentioned, #3361 changed this behavior to prefer a graceful exit. This **is** objectively better and will cause fewer bugs with resources left open when processes are force quit.

That said, I’m going to recategorize this one from a bug to an enhancement: we should make this behavior smarter. SIGINT should be able to interrupt in-process work and shouldn’t wait for the build to finish.

**Comment by ronaldroe** [comment]
Could we get some kind of output to know that it's working on it? At least that way people who haven't read this issue are sitting there banging CTRL + c over and over wondering why it isn't working. I ended up just killing my terminal the first couple times.

Something like `exiting. this may take a few moments`, or similar.

**Reply by cfjedimaster** [reply]
I'd just like to plus one this myself (yeah, I know I can click the button above, but wanted to add some commentary). I just assumed it was a shell issue (recently switched to OSX). It takes 30+ seconds to stop Eleventy now and there's no feedback on why.

**Reply by zachleat** [reply]
Upvote #4221, related to this!
